File: src/abn_combined/core/dedup.py
"""Deterministic transaction identity and duplicate handling (ported)."""

from __future__ import annotations

import hashlib
from typing import Any

from sqlalchemy.orm import Session

from .models import Transaction
from .utils import calculate_transaction_hash_components, normalize_category

DEFAULT_CURRENCY = "EUR"


def generate_transaction_id(trans: dict[str, Any]) -> str:
    """Generate a deterministic unique ID for a transaction.

    PayPal: ``account_paypal_transaction_id``. Wise: ``account_wise_transaction_id``.
    SEB: ``account_seb_voucher_id``. Otherwise ``account_date_amount_deschash``.
    """
    account = str(trans.get("accountNumber", ""))

    paypal_txn_id = trans.get("paypal_transaction_id")
    if paypal_txn_id:
        return f"{account}_{paypal_txn_id}"

    wise_txn_id = trans.get("wise_transaction_id")
    if wise_txn_id:
        return f"{account}_{wise_txn_id}"

    seb_voucher_id = trans.get("seb_voucher_id")
    if seb_voucher_id:
        return f"{account}_{seb_voucher_id}"

    date = str(trans.get("transactiondate", ""))
    amount = str(trans.get("amount", ""))
    description = str(trans.get("description", ""))
    desc_hash = hashlib.md5(description.encode("utf-8")).hexdigest()[:16]
    return f"{account}_{date}_{amount}_{desc_hash}"


def check_duplicates(
    db: Session, transactions: list[dict[str, Any]]
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    """Split incoming transactions into (new, duplicate) using deterministic ids.

    Duplicate = id already present in the DB, or a duplicate id within the batch.
    """
    incoming_ids: dict[str, dict[str, Any]] = {}
    duplicate_transactions: list[dict[str, Any]] = []

    for trans in transactions:
        trans_id = generate_transaction_id(trans)
        if trans_id in incoming_ids:
            # In-batch duplicate.
            duplicate_transactions.append(trans)
        else:
            incoming_ids[trans_id] = trans

    existing_ids: set[str] = set()
    if incoming_ids:
        existing_ids = {row[0] for row in db.query(Transaction.id).all()}

    new_transactions: list[dict[str, Any]] = []
    for trans_id, trans in incoming_ids.items():
        if trans_id in existing_ids:
            duplicate_transactions.append(trans)
        else:
            new_transactions.append(trans)

    return new_transactions, duplicate_transactions


def insert_transactions(db: Session, transactions: list[dict[str, Any]]) -> list[str]:
    """Insert transactions, returning the list of ids written.

    Uses ``db.merge`` so re-inserting an existing id is idempotent (exact duplicates
    are silently absorbed). Callers that need duplicate counts should call
    :func:`check_duplicates` first.
    """
    written: list[str] = []
    for trans in transactions:
        trans_id = generate_transaction_id(trans)
        transaction_hash = calculate_transaction_hash_components(
            date_value=trans.get("transactiondate"),
            description=trans.get("description"),
            amount=trans.get("amount"),
            account=trans.get("accountNumber"),
        )
        db_transaction = Transaction(
            id=trans_id,
            accountNumber=trans.get("accountNumber", ""),
            mutationcode=trans.get("mutationcode"),
            transactiondate=trans.get("transactiondate"),
            valuedate=trans.get("valuedate"),
            startsaldo=trans.get("startsaldo"),
            endsaldo=trans.get("endsaldo"),
            amount=trans.get("amount"),
            description=trans.get("description"),
            description_structured=trans.get("description_structured"),
            category=normalize_category(trans.get("category")) or trans.get("category"),
            manual_category=(
                normalize_category(trans.get("manual_category"))
                or trans.get("manual_category")
            ),
            tags=trans.get("tags"),
            manual_tags=trans.get("manual_tags"),
            categorization_source=trans.get("categorization_source"),
            currency=trans.get("currency", DEFAULT_CURRENCY),
            source_file=trans.get("source_file"),
            source_line=trans.get("source_line"),
            transaction_type_code=trans.get("transaction_type_code"),
            transaction_reference=trans.get("transaction_reference"),
            transaction_hash=transaction_hash,
        )
        db.merge(db_transaction)
        written.append(trans_id)
    db.commit()
    return written

File: ios/DVMFinanceKit/Sources/DVMFinanceKit/Core/Dedup.swift
import Foundation
import GRDB

/// Port of `core/dedup.py`.
///
/// `check_duplicates`/`insert_transactions` operate on lists of transaction
/// dicts in Python; here they operate on `[ParsedTransaction]`. `db` is a
/// GRDB `Database` (a connection inside a read/write closure), not the
/// `AppDatabase` wrapper — callers open the transaction via
/// `appDatabase.dbWriter.write { db in ... }`, matching how `Categorizer.swift`
/// takes `db: Database` too.
public enum Dedup {

    /// Errors that only arise from insert-time data that violates the
    /// desktop schema's `NOT NULL` constraints (`core/models.py`:
    /// `transactiondate`/`amount` are `nullable=False`). Real Phase D parsers
    /// always populate both fields for a parsed statement line; this only
    /// guards against malformed input reaching `insertTransactions`,
    /// mirroring the `IntegrityError` SQLAlchemy would raise for the same
    /// condition in the Python pipeline, just surfaced before we attempt the
    /// write rather than as a SQLite constraint failure.
    public enum DedupError: Error, Equatable {
        case missingTransactionDate
        case missingAmount
    }

    /// Port of `core/dedup.py:check_duplicates`.
    ///
    /// Splits incoming transactions into `(new, duplicates)` using
    /// deterministic ids (`TransactionID.generateTransactionID`). A
    /// duplicate is either a repeat id *within* `transactions` (first
    /// occurrence wins, matching Python's dict-insertion-order semantics) or
    /// an id that already exists in `db`.
    public static func checkDuplicates(
        db: Database,
        transactions: [ParsedTransaction]
    ) throws -> (new: [ParsedTransaction], duplicates: [ParsedTransaction]) {
        var incomingByID: [String: ParsedTransaction] = [:]
        var incomingOrder: [String] = []
        var duplicates: [ParsedTransaction] = []

        for transaction in transactions {
            let id = TransactionID.generateTransactionID(transaction)
            if incomingByID[id] != nil {
                // In-batch duplicate.
                duplicates.append(transaction)
            } else {
                incomingByID[id] = transaction
                incomingOrder.append(id)
            }
        }

        var existingIDs: Set<String> = []
        if !incomingByID.isEmpty {
            let idsRequest = TransactionRecord.select(Column("id"), as: String.self)
            existingIDs = try Set(idsRequest.fetchAll(db))
        }

        var newTransactions: [ParsedTransaction] = []
        newTransactions.reserveCapacity(incomingOrder.count)
        for id in incomingOrder {
            let transaction = incomingByID[id]!
            if existingIDs.contains(id) {
                duplicates.append(transaction)
            } else {
                newTransactions.append(transaction)
            }
        }

        return (newTransactions, duplicates)
    }

    /// Port of `core/dedup.py:insert_transactions`.
    ///
    /// Upserts (GRDB `save(_:)`, the analog of SQLAlchemy `Session.merge`)
    /// each transaction, computing its id and `transaction_hash`, applying
    /// `normalize_category` to `category`/`manual_category`, and defaulting
    /// `currency` to `"EUR"`. Returns the ids written, in input order.
    ///
    /// Idempotent: re-inserting an id already present overwrites that row
    /// (exact re-imports are silently absorbed) — callers that need
    /// duplicate *counts* must call `checkDuplicates` first, exactly as the
    /// Python docstring notes.
    @discardableResult
    public static func insertTransactions(
        db: Database,
        transactions: [ParsedTransaction]
    ) throws -> [String] {
        var written: [String] = []
        written.reserveCapacity(transactions.count)

        for transaction in transactions {
            guard let transactiondate = transaction.transactiondate else {
                throw DedupError.missingTransactionDate
            }
            guard let amount = transaction.amount else {
                throw DedupError.missingAmount
            }

            let id = TransactionID.generateTransactionID(transaction)
            let transactionHash = CoreNormalize.calculateTransactionHash(
                date: transaction.transactiondate,
                description: transaction.description,
                amount: CoreNormalize.HashAmountInput(transaction.amount),
                account: transaction.accountNumber
            )

            // Port of Python's `normalize_category(x) or x`: if normalization
            // yields `nil` (blank/whitespace-only input, or `x` was already
            // `nil`), fall back to the raw value rather than dropping it —
            // this only differs from `normalizeCategory(x)` when `x` itself
            // is a non-nil string that normalizes to `nil` (e.g. all-comma
            // input), in which case Python keeps the *original* unnormalized
            // string.
            let category = CoreNormalize.normalizeCategory(transaction.category) ?? transaction.category
            let manualCategory = CoreNormalize.normalizeCategory(transaction.manualCategory)
                ?? transaction.manualCategory

            let record = TransactionRecord(
                id: id,
                accountNumber: transaction.accountNumber,
                mutationcode: transaction.mutationcode,
                transactiondate: transactiondate,
                valuedate: transaction.valuedate,
                startsaldo: transaction.startsaldo.map { Decimal($0) },
                endsaldo: transaction.endsaldo.map { Decimal($0) },
                amount: Decimal(amount),
                description: transaction.description,
                descriptionStructured: transaction.descriptionStructured,
                category: category,
                manualCategory: manualCategory,
                tags: transaction.tags,
                manualTags: transaction.manualTags,
                categorizationSource: transaction.categorizationSource,
                currency: transaction.currency ?? "EUR",
                sourceFile: transaction.sourceFile,
                sourceLine: transaction.sourceLine,
                transactionTypeCode: transaction.transactionTypeCode,
                transactionReference: transaction.transactionReference,
                transactionHash: transactionHash
            )

            try record.save(db)
            written.append(id)
        }

        return written
    }
}
